ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/StatementExecutor.java

```diff
@@ -290,34 +267,40 @@ private void startQuery(
 
     if (QueryCapacityUtil.exceedsPersistentQueryCapacity(ksqlEngine, mergedConfig,1)) {
       QueryCapacityUtil.throwTooManyActivePersistentQueriesException(
-          ksqlEngine, mergedConfig, queryString);
+          ksqlEngine, mergedConfig, statement.getStatementText());
     }
 
     final QueryMetadata queryMetadata = ksqlEngine.execute(
-        queryString,
+        statement,
         mergedConfig,
         command.getOverwriteProperties()
-    ).get(0);
+    ).orElseThrow(() -> new KsqlException("Statement did not return a query"));
 
-    if (queryMetadata instanceof PersistentQueryMetadata) {
-      final PersistentQueryMetadata persistentQueryMd = (PersistentQueryMetadata) queryMetadata;
-      if (mode == Mode.EXECUTE) {
-        persistentQueryMd.start();
-      }
-    } else {
+    if (!(queryMetadata instanceof PersistentQueryMetadata)) {
       throw new KsqlException(String.format(
           "Unexpected query metadata type: %s; was expecting %s",
           queryMetadata.getClass().getCanonicalName(),
           PersistentQueryMetadata.class.getCanonicalName()
       ));
     }
+
+    final PersistentQueryMetadata persistentQueryMd = (PersistentQueryMetadata) queryMetadata;
+    if (mode == Mode.EXECUTE) {
+      persistentQueryMd.start();
+    }
   }
 
-  private void terminateQuery(final TerminateQuery terminateQuery) {
-    final QueryId queryId = terminateQuery.getQueryId();
-    if (!ksqlEngine.terminateQuery(queryId)) {
```

Comment:
It looks like we have 2 ways to clean up queries now - KsqlEngine::terminateQuery, and QueryMetadata::close. We should clean up the former. I'll file an issue.

Reply:
Don't bother. I think that's a valid point and so I've removed the terminateQuery method.  You can now terminate a query by getting it, (i.e. via getPersistentQuery(queryId)), and then closing it.